### ISSUE TITLE
Cordova Support

### ISSUE THREAD

**Issue by sandorfr** [mixpanel/mixpanel-js#24]
Do you have plans regarding cordova support?

There are two main concerns when it comes to cordova support using mixpanel-js:
1. The current implementation is cookie based
2. Offline support

There is a simple [workaround](https://github.com/sandorfr/mixpanel-js/compare/master...sandorfr:cordova) for the first one which consists of replacing the current cookie implementation by a localstorage based implementation. Currently this is dirty but I can work on a pull request if you are interested. 

I don't have worked yet on the second issue but I think we can build a local storage solution too. I think this would be awesome for cordova but this would help people who build websites with offline support too.

**Comment by tdumitrescu** [comment]
Hi, thanks so much for pointing out the issues and potential solutions. Offline support and alternatives to cookie-based tracking would be awesome. Unfortunately there's nothing concrete currently on the roadmap, so pull requests are probably the quickest path to seeing the features you need. A few notes:
- if you do implement both of these, please keep them in separate pull requests
- for localStorage vs cookie stores, maintain backwards compatibility by offering localStorage as a config option and keeping cookie as the default; so you would init the lib somewhat like: `mixpanel.init(TOKEN, {storage: 'localStorage'});`
- for offline support, you'll want to queue up requests and then send batched updates (https://mixpanel.com/help/reference/http#batch-requests) of max 40 events at a time. the code of our [ios lib](https://github.com/mixpanel/mixpanel-iphone) offers a good model.

Let us know if you have more questions!

**Comment by sandorfr** [comment]
Hi, yes there is one thing. I'm struggling running tests do you have any guidance on that?
My issue seems to be related with cookie_included getJson calls (which are failing with a 404).

**Comment by tdumitrescu** [comment]
Apologies for that; the tests do rely on an external server resource which wasn't included when the lib was open-sourced. I've been meaning to add a little Node server to the repo to handle that. The Django code for it looks basically like:
```python
def test_cookie_included(request, cookie_name):
    return json_response(request, 1 if cookie_name in request.COOKIES else 0)
```
so the response is 1 when the cookie has been set and 0 otherwise. Does that help?

**Comment by sandorfr** [comment]
It does help :) Actually I wrote a basic nodejs/express server which I will update with using that information. I can submit another PR for that too if you want.

**Comment by tdumitrescu** [comment]
That would be very welcome!

**Comment by sandorfr** [comment]
Just submitted a PR about the test server. 
I still have some work to do about localStorage to make it clean but I'm on it.

**Comment by sandorfr** [comment]
I've added a pull request with the localStorage alternative to cookies.

**Comment by KeKs0r** [comment]
@sandorfr I see that you already solved one of the two problems for good cordova support. Did you also tackle the queue topic?

**Comment by sandorfr** [comment]
@KeKs0r I have not yet started it.

**Comment by mobidev111** [comment]
here's a queue implementation:
http://meltuhamy.com/dev/offline-resilient-mixpanel-tracking-for-ionic-without-a-cordova-plugin/
https://github.com/meltuhamy/belfastsalah/blob/master/www/js/svc/mixpanel.js

@sandorfr @tdumitrescu any chance to take on this in the foreseeable future?
similarly to what you did with the localstorage support?

**Comment by mobidev111** [comment]
alternative approach here:
https://github.com/skotzko/mixpanel-event-queue

**Comment by tdumitrescu** [comment]
The best way to push a feature like this forward is with a pull request, otherwise it's largely a matter of us prioritizing the request against a whole pile of other internal development work, which means a longer wait. I'm happy to discuss ideas if you want to validate an approach before submitting code.

**Comment by sandorfr** [comment]
Hi @mobidev111, I don't MEQ is going to be of any help when it comes to offline support because this is using in Memory storage. 

I was considering implementing a queue on top of IndexedDb and then batch the events. Since reinventing the wheel is not going to help I think it might be interresting to take a look at this  : https://github.com/orthecreedence/hustle.

That being said there is some work and a lot of testing to be done to get something which could be merged here, and I'm also prioritizing things on my side. So unless someone makes it a priority for me I'm very short on time ATM.

**Comment by jskrepnek** [comment]
To add another consideration for Cordova support, it looks like in all scenarios mixpanel will default the protocol to 'http' rather than 'https', since the protocol is actually 'file'.

See:
https://github.com/mixpanel/mixpanel-js/blob/40c158c0030a65df0c203a2817f917757af09666/src/mixpanel-core.js#L72

**Comment by sandorfr** [comment]
@jskrepnek interesting, and it looks like this is not configurable yet.

**Comment by jskrepnek** [comment]
@sandorfr I've successfully patched it in place, but that's far from ideal.

**Comment by tdumitrescu** [comment]
It is currently configurable to the extent that you can specify the entire api host string in the config, though it's certainly not ideal:
```javascript
mixpanel.init('MY TOKEN', {
  api_host: 'https://api.mixpanel.com'
});
```
What would be the most useful means of configuration for you?

**Comment by jskrepnek** [comment]
@tdumitrescu That's great, thanks.  I'd suggest adding that configuration option to the Javascript Full API Reference.

**Comment by jskrepnek** [comment]
Would it make sense to make 'https' the default, rather than 'http'?  That was cases like cordova where the protocol is 'file' would default the safe way.

**Comment by tdumitrescu** [comment]
Good call, will need to add it [here](https://github.com/mixpanel/mixpanel-js/blob/master/src/mixpanel-core.js#L1448-L1489).

**Comment by tdumitrescu** [comment]
Yeah, we'll have to think about potential implications for backwards-compatibility with switching the default, though it seems safe to me off the top of my head.

**Comment by jfbloom22** [comment]
I believe this issue can be closed.  I am using Mixpanel in several Cordova apps at this point.  If you are using bower and your app goes offline: notice the section titled:  [Using Bower to load the entire library](https://github.com/mixpanel/mixpanel-js#using-bower-to-load-the-entire-library) which explains how to load Mixpanel without the CDN.  This will allow it to pass the CSP for Chrome:extensions as well.
Side note: the instructions are slightly wrong.  `mixpanel.js` does not exist.  It should be `mixpanel.min.js`.
